eva/experimental/ray/executor/exchange_executor.py

```diff
@@ -32,74 +33,70 @@ def __init__(self):
-    """
-    Applies predicates to filter the frames which satisfy the condition
-    Arguments:
-        node (AbstractPlan): The SequentialScanPlan
-
-    """
-
     def __init__(self, node: ExchangePlan):
+        self.inner_plan = node.inner_plan
         self.parallelism = node.parallelism
-        self.ray_conf = node.ray_conf
+        self.ray_pull_env_conf_dict = node.ray_pull_env_conf_dict
+        self.ray_parallel_env_conf_dict = node.ray_parallel_env_conf_dict
         super().__init__(node)
 
-    def exec(self, is_top=True) -> Iterator[Batch]:
-        assert (
-            len(self.children) == 1
-        ), "Exchange executor does not support children != 1"
+    def build_inner_executor(self, inner_executor):
+        self.inner_executor = inner_executor
+        self.inner_executor.children = [QueueReaderExecutor()]
 
-        # Find the exchange exector below the tree
-        curr_exec = self
-        input_queues = []
-        while len(curr_exec.children) > 0 and not isinstance(
-            curr_exec.children[0], ExchangeExecutor
-        ):
-            curr_exec = curr_exec.children[0]
+    def exec(self) -> Iterator[Batch]:
+        input_queue = Queue(maxsize=100)
+        output_queue = Queue(maxsize=100)
 
-        if len(curr_exec.children) > 0:
-            iq = yield from curr_exec.children[0].exec(is_top=False)
-            input_queues.append(iq)
-            queue_exec = QueueReaderExecutor()
-            curr_exec.children = [queue_exec]
-            # curr_exec.append_child(queue_exec)
+        # Pull data from child executor
+        assert (
+            len(self.children) == 1
+        ), "Exchange currently only supports parallelization of node with only one child"
+        ray_pull_task = ray_pull.remote(
+            self.ray_pull_env_conf_dict,
+            self.children[0],
+            input_queue,
+        )
 
-        output_queue = Queue(maxsize=100)
-        ray_task = []
-        for _ in range(self.parallelism):
-            ray_task.append(
-                ray_stage.options(**self.ray_conf).remote(
-                    self.children[0], input_queues, [output_queue]
+        # Parallel the inner executor.
+        ray_parallel_task_list = []
+        for i in range(self.parallelism):
+            ray_parallel_task_list.append(
+                ray_parallel.remote(
+                    self.ray_parallel_env_conf_dict[i],
+                    self.inner_executor,
+                    input_queue,
+                    output_queue,
                 )
             )
-        ray_stage_wait_and_alert.remote(ray_task, [output_queue])
-        while is_top:
+
+        ray_wait_and_alert.remote([ray_pull_task], input_queue)
+        ray_wait_and_alert.remote(ray_parallel_task_list, output_queue)
+
+        while True:
             res = output_queue.get(block=True)
             if res is StageCompleteSignal:
                 break
             elif isinstance(res, ExecutorError):
```

Comment:
Maybe we can find a better way to garbage collect the ray actors and handle errors. I will think about it. Let us know if you have any thoughts.

Reply:
I see. Do you have some error cases example in mind, that may not be cleanly caught by current implementations?

Reply:
I see. I added a test case for error ray exception. The current print consists output from the main process and all remote processes. I remove the error logging of remote processes (which is just duplicate of the main process), so now it only prints one line of clean error. E.g., 

> 05-18-2023 10:32:02 ERROR [plan_executor:plan_executor.py:execute_plan:0177] The dataset file could not be found. Please verify that the file exists in the specified path.

And, I think the main process always returns one error message because it immediately stops when one ExecutorError is raised by any of remote processes.